src/Reconciler.js

```diff
@@ -2,7 +2,8 @@
 
 import Reconciler, { type HostConfig } from 'react-reconciler';
 import getOwnerDocument from 'dom-helpers/ownerDocument';
-
+import { SSRHydrationDev } from './SSRHydrationDev.js';
+import { SSRHydrationProd } from './SSRHydrationProd.js';
```

Comment:
we should add a comment or something here to make explicity that one _must_ be treeshaked, we should also maybe add some sort of test to ensure that? If one potential global side effect is introduced in any of these files the DCE will fail...

Reply:
> If one potential global side effect is introduced in any of these files the DCE will fail...

I dont understand why the DCE would fail due to a side effect? Does Rollup have a way to detect side effects without evaluating the code? If not, then we can be sure of seeing the following

```
if (__DEV__) {
 // A
} else {
// B
}
```

reduced to 


```
{
 // B
}
```

in production

Reply:
The comment sounds helpful I agree.

Reply:
it's specific to treeshacking as an optimization:

```
if (__DEV__) {
 // A
} else {
// B
}
```
That bit won't change, but whats happing is that in prod `A` is removed, and there are no references to the DEV import anymore so Rollup removes it. _But_ if the Dev file (or any file it imports) does something that looks like its not a pure function, e.g. sets a `global` variable, or references certain objects. Rollup can't guarantee that its not doing any work and will leave the file imported in the top

Reply:
> Rollup can't guarantee that its not doing any work and will leave the file imported in the top

Can you verify this from the bundle? When I checked only one of the module was imported.

Reply:
In dev build only SSRHydrationDev was present and in the minified build SSRHydrationProd was present.

Reply:
sorry i'm saying it broken now. I saying it's dangerous because it could break quietly and without obvious cause in the future, as it is this is working great right now

Reply:
Hey, I'm out of ideas for this problem. Can we make do with just a comment for now?

Reply:
ya definitely 👍

Reply:
> Caution! One one of the following modules is supposed to be imported. Avoid side effects in them.

Will this do for the comment?